### PR TITLE
refactor http request as a stand alone class

### PR DESCRIPTION
**Pull request by idohalevi**
Took out all the logic related to the HTTP request - will serve us better in jenkins plugin

### CODE REVIEW

File: logzio-sender/src/main/java/io/logz/sender/HttpsSyncRequest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import io.logz.sender.exceptions.LogzioServerErrorException;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import java.net.HttpURLConnection;
+
+public class HttpsSyncRequest {
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/HttpsSyncRequest.java]: It's not a request object, but more a Sender. Maybe `HttpsSyncSender`?

File: logzio-sender/src/main/java/io/logz/sender/LogzioSender.java [hunk 2]
```diff
@@ -265,7 +245,7 @@ private int sizeInBytes(List<FormattedLogMessage> logMessages) {
 
     private byte[] toNewLineSeparatedByteArray(List<FormattedLogMessage> messages) {
         try (ByteArrayOutputStream byteOutputStream = new ByteArrayOutputStream(sizeInBytes(messages));
-             OutputStream os = compressRequests ? new GZIPOutputStream(byteOutputStream) : byteOutputStream) {
+             OutputStream os = httpsRequestConfiguration.isCompressRequests() ? new GZIPOutputStream(byteOutputStream) : byteOutputStream) {
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/LogzioSender.java]: This should be inside the request sender I believe.

Comment (idohalevi) [reply logzio-sender/src/main/java/io/logz/sender/LogzioSender.java]: @asafm 
I feel like this should be in the java sender and not at the sync client, I want the client to be as simple as possible - getting msg and configuration and that's it. Whoever uses it should make sure the msg is byte[]. What do you think?

File: logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                        new URL(url + "/?token=" + logzioToken) :
+                        new URL(url + "/?token=" + logzioToken + "&type=" + logzioType);
+        }
+
+        public HttpsRequestConfiguration build() throws LogzioParameterErrorException {
+            URL url;
+            try {
+                url = createURL(logzioListenerUrl);
+            } catch (MalformedURLException e){
+                throw new LogzioParameterErrorException("logzioUrl="+logzioListenerUrl+" token="+logzioToken+" type="+logzioType, "For some reason could not initialize URL. Cant recover.." + e);
```
Comment (MarinaRazumovsky) [inline logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java]: add logging , not only exception

File: logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+                    reporter.warning("Could not send log to logz.io, retry (" + currTry + "/" + configuration.getMaxRetriesAttempts()+ ")");
+                    reporter.warning("Sleeping for " + currentRetrySleep + " ms and will try again.");
+                    Thread.sleep(currentRetrySleep);
+                    currentRetrySleep *= 2;
+                }
+            }
+
+        } catch (InterruptedException e) {
+            reporter.error("Got interrupted exception");
```
Comment (MarinaRazumovsky) [inline logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java]: Are you sure that it is error loglevel ?

File: logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java [hunk 5]
```diff
@@ -2,12 +2,11 @@
 
 import io.logz.sender.exceptions.LogzioServerErrorException;
 
-import java.io.BufferedReader;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
+import java.io.*;
```
Comment (MarinaRazumovsky) [inline logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java]: not *

File: logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java [hunk 6]
```diff
@@ -56,7 +56,7 @@ public boolean isCompressRequests() {
     private final boolean compressRequests;
 
 
-    public HttpsRequestConfiguration(String logzioToken,
+    private HttpsRequestConfiguration(String logzioToken,
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java]: If this breaks backward compatibility - you need to mention that in the release notes

File: logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java [hunk 7]
```diff
@@ -83,6 +83,7 @@ public HttpsRequestConfiguration(String logzioToken,
         private String logzioListenerUrl = "https://listener.logz.io:8071";
         private String logzioToken;
         private boolean compressRequests = false;
+        private SenderStatusReporter reporter;
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java]: A configuration object (i.e. a data structure) should not contain objects to act upon like a SenderStatusReporter

File: logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java [hunk 8]
```diff
@@ -131,6 +132,11 @@ public Builder setCompressRequests(boolean compressRequests) {
             return this;
         }
 
+        public Builder setReporter(SenderStatusReporter reporter) {
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/HttpsRequestConfiguration.java]: Why not make this configuration object immutable?

File: logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java [hunk 9]
```diff
@@ -35,8 +40,28 @@ private boolean shouldRetry(int statusCode) {
         return shouldRetry;
     }
 
-    public void sendToLogzio(byte[] lineSeparatedPayload) throws LogzioServerErrorException {
+    private byte[] toNewLineSeparatedByteArray(List<FormattedLogMessage> messages) {
+        try (ByteArrayOutputStream byteOutputStream = new ByteArrayOutputStream(sizeInBytes(messages));
+             OutputStream os = configuration.isCompressRequests() ? new GZIPOutputStream(byteOutputStream) : byteOutputStream) {
+            for (FormattedLogMessage currMessage : messages) os.write(currMessage.getMessage());
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java]: Better to break it with {}, easier to read the for loop

File: logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java [hunk 10]
```diff
@@ -35,8 +40,28 @@ private boolean shouldRetry(int statusCode) {
...
+            // Need close before return for gzip compression, The stream only knows to compress and write the last bytes when you tell it to close
+            os.close();
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java]: It's not closed upon exception. Maybe through try-with-resources?

File: logzio-sender/src/main/java/io/logz/sender/LogzioSender.java [hunk 11]
```diff
@@ -39,16 +35,15 @@
...
+    private final HttpsSyncSender httpsSyncSender;
 
     private LogzioSender(String logzioToken, String logzioType, int drainTimeout, int fsPercentThreshold, File bufferDir,
                          String logzioUrl, int socketTimeout, int connectTimeout, boolean debug,
                          SenderStatusReporter reporter, ScheduledExecutorService tasksExecutor,
                          int gcPersistedQueueFilesIntervalSeconds, boolean compressRequests)
             throws  LogzioParameterErrorException {
 
-        httpsRequestConfiguration = HttpsRequestConfiguration
+        HttpsRequestConfiguration httpsRequestConfiguration = HttpsRequestConfiguration
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/LogzioSender.java]: The build() should be the last step. Usually it's xxBuilder.newBuilder().setY().setZ().build() --> xx

File: logzio-sender/src/main/java/io/logz/sender/LogzioSender.java [hunk 12]
```diff
@@ -198,11 +193,13 @@ private boolean isEnoughDiskSpace() {
 
     private List<FormattedLogMessage> dequeueUpToMaxBatchSize() {
         List<FormattedLogMessage> logsList = new ArrayList<>();
+        int totalSize = 0;
```
Comment (asafm) [inline logzio-sender/src/main/java/io/logz/sender/LogzioSender.java]: always work with long with sizes, for safety reasons.

**Pull request merged**